android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java

```diff
@@ -785,6 +779,90 @@ public boolean accept(File pathname) {
     return list;
   }
 
+  /**
+   * Loop through the installed keyboards list:
+   * If the keyboard exists in cloud/ and packages/, migrate keyboard entry to packages/
+   * Then remove the keyboard files in cloud/
+   * @param context
+   */
+  public static void migrateCloudKeyboards(Context context) {
+    boolean keyboardMigrated = false;
+    for(int i=0; i<KeyboardController.getInstance().get().size(); i++) {
+      // Cloud keyboard info
+      Keyboard k = KeyboardController.getInstance().getKeyboardInfo(i);
+      if (k.getPackageID().equals(KMManager.KMDefault_UndefinedPackageID)) {
+        String keyboardID = k.getKeyboardID();
+        // We'll want to preserve language fields
+        String languageID = k.getLanguageID();
+        String languageName = k.getLanguageName();
+
+        // Look for a non-cloud packageID. See if there's one from the updateKMP field,
+        // otherwise fall back to keyboardID
+        String packageID = keyboardID;
+        if (k.getUpdateKMP() != null && !k.getUpdateKMP().isEmpty()) {
+          packageID = FileUtils.getFilename(k.getUpdateKMP()).replace(FileUtils.KEYMANPACKAGE, "");
+        }
+
+        // See if there's a matching keyboard of (package ID, keyboardID, ignoring language ID)
+        int keyboardIndex = KeyboardController.getInstance().getKeyboardIndex(
+          packageID, keyboardID, "");
+        if (keyboardIndex != KeyboardController.INDEX_NOT_FOUND) {
+          Keyboard keyboardFromPackage = KeyboardController.getInstance().getKeyboardInfo(keyboardIndex);
+          // Create a copy of keyboardFromPackage, updating the language ID and language Name
+          Keyboard migratedKeyboard = new Keyboard(keyboardFromPackage);
+          migratedKeyboard.setLanguage(languageID, languageName);
+          KeyboardController.getInstance().set(i, migratedKeyboard);
+
+          // Remove the cloud keyboard files
+          removeCloudKeyboard(keyboardID);
+          keyboardMigrated = true;
+        }
+      }
+    }
+
+    if (keyboardMigrated) {
+      KeyboardController.getInstance().save(context);
+    }
+  }
+
+  /**
+   * Remove all the keyboard files in the "cloud" directory matching a pattern
+   * [keyboard ID]-[version].js
+   *
+   * @param keyboardID String of the keyboard ID
+   */
+  public static void removeCloudKeyboard(final String keyboardID) {
+    String path = getCloudDir();
+    File dir = new File(path);
+
+    FileFilter keyboardFilter = new FileFilter() {
+      @Override
+      /**
+       * Filter for JS keyboards that match [keyboardID]-[version].js
+       */
+      public boolean accept(File pathname) {
+        String name = pathname.getName();
+
+        String patternStr = String.format("^(%s)-([0-9.]+)(\\.js)$", keyboardID);
+        Pattern pattern = Pattern.compile(patternStr);
+        Matcher matcher = pattern.matcher(name);
+        if (matcher.matches() && (matcher.group(1) != null)) {
```

Comment:
This should work because we are fairly strict on what we allow for web keyboards -- they have to be valid Javascript identifiers which pretty much rules out any regex special chars which could otherwise sneak in through `keyboardID`. But as a general rule, injecting external strings into a regex is a recipe for future headaches. It's almost always safer to do something like this:

```suggestion
        Pattern pattern = Pattern.compile("^([A-Za-z0-9_]+)-([0-9.]+)(\\.js)$");
        Matcher matcher = pattern.matcher(name);
        if (matcher.matches() && matcher.group(1) == keyboardID && (matcher.group(2) != null)) {
```